handler/src/main/java/io/netty/handler/ssl/ConscryptAlpnSslEngine.java

```diff
@@ -0,0 +1,181 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.lang.Math.min;
+
+import io.netty.handler.ssl.JdkApplicationProtocolNegotiator.ProtocolSelectionListener;
+import io.netty.handler.ssl.JdkApplicationProtocolNegotiator.ProtocolSelector;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLHandshakeException;
+import org.conscrypt.Conscrypt;
+import org.conscrypt.HandshakeListener;
+
+/**
+ * A {@link JdkSslEngine} that uses the Conscrypt provider or SSL with ALPN.
+ */
+abstract class ConscryptAlpnSslEngine extends JdkSslEngine {
+    private static final Class<?> ENGINES_CLASS = getEnginesClass();
+
+    /**
+     * Indicates whether or not conscrypt is available on the current system.
+     */
+    static boolean isAvailable() {
+        return ENGINES_CLASS != null;
+    }
+
+    static boolean isEngineSupported(SSLEngine engine) {
+        return isAvailable() && isConscryptEngine(engine, ENGINES_CLASS);
+    }
+
+    static ConscryptAlpnSslEngine newClientEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ClientEngine(engine, applicationNegotiator);
+    }
+
+    static ConscryptAlpnSslEngine newServerEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ServerEngine(engine, applicationNegotiator);
+    }
+
+    private ConscryptAlpnSslEngine(SSLEngine engine, List<String> protocols) {
+        super(engine);
+
+        // Set the list of supported ALPN protocols on the engine.
+        Conscrypt.Engines.setAlpnProtocols(engine, protocols.toArray(new String[protocols.size()]));
+    }
+
+    /**
+     * Calculates the maximum size of the encrypted output buffer required to wrap the given plaintext bytes. Assumes
+     * as a worst case that there is one TLS record per buffer.
+     *
+     * @param plaintextBytes the number of plaintext bytes to be wrapped.
+     * @param numBuffers the number of buffers that the plaintext bytes are spread across.
+     * @return the maximum size of the encrypted output buffer required for the wrap operation.
+     */
+    final int calculateOutNetBufSize(int plaintextBytes, int numBuffers) {
+        // Assuming a max of one frame per component in a composite buffer.
+        long maxOverhead = (long) Conscrypt.Engines.maxSealOverhead(getWrappedEngine()) * numBuffers;
+        return (int) min(Integer.MAX_VALUE, plaintextBytes + maxOverhead);
+    }
+
+    final SSLEngineResult unwrap(ByteBuffer[] srcs, ByteBuffer[] dests) throws SSLException {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)